### ISSUE TITLE
Car colours

### ISSUE THREAD

**Issue by Y-Less** [Open-GTO/sa-mp-fixes#36]
This is something that I have wanted to do for a while.  When you spawn a car with colour `-1`, one is randomly selected; however, it is not synced.  This is different to using `random(127)`, since only valid colours for the vehicle are used. Speaking of which, could anyone send me a copy of `carcols.dat` please?

**Comment by Y-Less** [comment]
Solving it is easy - just keep the carcols.dat data server-side.  In fact, there is no reason not to load that information from a file directly, other than the fact that it adds the first dependency ever.

**Comment by IstuntmanI** [comment]
Well, there's already an include on forums for that: http://forum.sa-mp.com/showthread.php?t=339088 . There's no need for a separate file.

**Comment by Y-Less** [comment]
There's actually at least two, the earlier one is here:

http://forum.sa-mp.com/showthread.php?t=235398

And has a comment from me from 2011 about fixing `-1` server-side.  When I say I've wanted to do it for a while - I mean it!

But yes, pulling that fix in to here would remove the need for a separate file.  And I realised that the `carcols.dat` file could also be made optional - just include the defaults in the script and if they don't have a custom version there's no need for loading it.

**Comment by rt-2** [comment]
Why not just hooking the spawning function, detecting if color is -1 and if so replace -1 by random(127).
I don't understand the need for ``carcols.dat``, sorry if I missed something.
rt-2

**Comment by Y-Less** [comment]
Because `-1` is not the same as `random(127)`.  There are a list of valid colour combinations for each vehicle in `carcols.dat`, like this:

```
sparrow, 1,3
speeder, 1,3, 1,5, 1,16, 1,22, 1,35, 1,44, 1,53, 1,57
supergt, 3,3, 6,6, 7,7, 52,52, 76,76
```

So the sparrow can only ever (in the original game) have one colour pair.  The supergt always has both colours the same, and so on.  Doing `random(127)` will just end up with all sorts of strange combinations, and anything will go.  `-1` pulls the data from `carcols.dat` and only selects valid combinations from that file.

**Comment by Y-Less** [comment]
Note that I'm just trying to list almost everything that could be fixed, so there's a reference somewhere.  Some may never by included in this include.  Although I've already written some of this code, just not the loading.

**Comment by rt-2** [comment]
Ok now I understand the problem.
Thank you

**Comment by WoutProvost** [comment]
If it is of any use, I've written some documentation on this fix a long time ago:
* Using fixed vehicle colours is synced between players, however using -1 as a vehicle colour is not.
* -1 selects a random vehicle colour of possible known singleplayer values. Therefore it is not completely random with values like 0 to 127.
* The singleplayer values are always a pair of 2 (1 exception, see below), of which the first value is the always the first colour and the second value is always the second colour.
* When no know singplayer values are present, a single value of -1 is used.
* When both to be assigned vehicle colours are -1, a random pair will be chosen. The first colour will be the first value, the second colour the second value.
* When only the first to be assigned vehicle colour is -1, a random pair is still chosen, but only the first value will be used. The second colour is fixed.
* When only the second to be assigned vehicle colour is -1, a random pair is still chosen, but only the second value will be used. The first colour is fixed.
* In other words, the first value can not be used for the second colour and vice versa.
* We can sync these random colours with ChangeVehicleColor.
* Upon spawning, the -1 colours are assigned and synced with ChangeVehicleColor.
* Upon respawning, the -1 colours are chosen again and synced with ChangeVehicleColor.
* If only one colour was -1 upon first spawning, only that colour will be chosen again upon respawning. The other colour remains fixed.

**Comment by WoutProvost** [comment]
Can be extended to (Player)TextDrawSetPreviewVehCol as well.

**Comment by WoutProvost** [comment]
So, what's the status on this? I tried messaging @Y-Less but, didn't get an answer. My current implementation can be found [here](https://github.com/WoutProvost/samp-util/blob/master/main/vehicle.inc#L107).

**Comment by Y-Less** [comment]
I've done this for vehicles.  I'll add it to `TextDrawSetPreviewVehCol`, but there's no point adding it to `PlayerTextDrawSetPreviewVehCol` since that's only seen by one player - it is automatically synced.
